src/lib.rs

```diff
@@ -42,232 +48,290 @@ struct FilterConfig {
+struct TokenResponse {
+    #[serde(default)]
+    error: String,
+    #[serde(default)]
+    error_description: String,
+    #[serde(default)]
+    id_token: String,
+    #[serde(default)]
+    expires_in: i64
+}
+
+#[derive(Serialize, Deserialize)]
+struct Claims {
+    nonce: String,
+    iss: String,
+    proto: String,
+    path: String,
+    redirect_uri: String
+}
+
+fn default_oidc_cookie_name() -> String {
+    "oidcToken".to_owned()
+}
+
+fn default_target_header_name() -> String {
+    "authorization".to_owned()
+}
+
 impl OIDCFilter {
-    fn is_authorized(&self) -> bool {
+    fn get_header(&self, name: &str) -> String {
         let headers = self.get_http_request_headers();
         for (key,_value) in headers.iter() {
-            if key.to_lowercase().trim() == "authorization" {
-                return true;
+            if key.to_lowercase().trim() == name {
+                return _value.to_owned();
             }
         }
-        return false;
+        return "".to_owned();
     }
 
     fn get_code(&self) -> String {
         let path = self.get_http_request_header(":path").unwrap();
-        let path_parts: Vec<_> = path.split("?").collect();
-        if path_parts.len() < 2 {
-            return "".to_string()
-        }
-        let query = path_parts[1].to_string();
+        let query_offset = path.find("?").unwrap_or(0) + 1;
+        if query_offset == 1 { return "".to_owned(); }
+
+        let query = &path[query_offset..path.len()];
         let encoded = form_urlencoded::parse(query.as_bytes());
         for (k, v) in encoded {
             if k == "code" {
                 return v.to_owned().to_string();
             }
         }
-        return "".to_string();
+        return "".to_owned();
     }
 
     fn get_cookie(&self, name: &str) -> String {
         let headers = self.get_http_request_headers();
-        let mut cookies = "";
         for (key,value) in headers.iter() {
             if key.to_lowercase().trim() == "cookie" {
-                cookies = value;
-            }
-        }
-        let assignments: Vec<_> = cookies.split(";").collect();
-        for assignment in assignments {
-            let kvpair: Vec<_> = assignment.split("=").collect();
-            if kvpair[0] == name {
-                return kvpair[1].to_owned();
+                let cookies: Vec<_> = value.split(";").collect();
+                for cookie_string in cookies {
+                    let cookie_name_end = cookie_string.find('=').unwrap_or(0);
+                    let cookie_name = &cookie_string[0..cookie_name_end];
+                    if cookie_name.trim() == name {
+                        return cookie_string[(cookie_name_end + 1)..cookie_string.len()].to_owned();
+                    }
+                }
             }
         }
         return "".to_owned()
     }
 
-    fn get_redirect_uri(&self, current_uri: &str) -> String {
+    fn get_callback_url(&self, host: &str, redirect_path: &str) -> String {
```

Comment:
we should consider adding `proto` as a parameter and moving this out of this class into a separate file. we can do this in a follow-up though, just taking mental notes here